### PR TITLE
Remaining solidity native errors

### PR DESCRIPTION
**Pull request by keirongulrajani**
## Scope

- Solidity native errors for:
  - Circuit breaker
  - Crypto Library
  - CustomEnumerableMapsErrors
  - Ethdkg + accusations + phases
  - Governance
  - Immuteable auth
  - parser libs: Bclaims/merkle proof/Pclaims/r cert/ r claims/ tx in pre image
  - snapshots
  - staking nft
  - validator pool
  - Merkle library

- Updated immutable auth to auto generate solidity native errors for generated immutable contracts

- Installed updated hardhat lib for asserting new custom errors in tests

- updated bindings

## Why?

[MP-591](https://madhive.atlassian.net/browse/MP-591)

## Todos

If any, what are the follow-up tasks required other than merging this PR? Have they been arranged?

- [ ] ???

### CODE REVIEW

File: bridge/contracts/BToken.sol [hunk 1]
```diff
@@ -501,6 +501,7 @@ contract BToken is
         if (totalSupply_ < numBTK_) {
             revert BTokenErrors.BurnAmountExceedsSupply(numBTK_, totalSupply_);
         }
+
```
Comment (vtleonardo) [inline bridge/contracts/BToken.sol:504]: Can we remove this file from this PR?

File: bridge/contracts/ETHDKG.sol [hunk 2]
```diff
@@ -229,15 +227,20 @@ contract ETHDKG is
...
-            validatorsAccounts_.length == validatorIndexes_.length &&
-                validatorsAccounts_.length == validatorShares_.length,
-            string(abi.encodePacked(ETHDKGErrorCodes.ETHDKG_MIGRATION_INPUT_DATA_MISMATCH))
-        );
+        if (nonce != 0) {
+            revert ETHDKGErrors.MigrationRequiresZeroNonce(nonce);
+        }
+
+        if (
+            !(validatorsAccounts_.length == validatorIndexes_.length &&
```
Comment (vtleonardo) [inline bridge/contracts/ETHDKG.sol:235]: Maybe reverting this for better readability? 
```
validatorsAccounts_.length != validatorIndexes_.length || validatorsAccounts_.length != validatorShares_.length
```

File: bridge/contracts/Snapshots.sol [hunk 3]
```diff
@@ -157,25 +157,25 @@ contract Snapshots is Initializable, SnapshotsStorage, ISnapshots {
...
-                string(abi.encodePacked(SnapshotsErrorCodes.SNAPSHOT_MIGRATION_NOT_ALLOWED))
-            );
-            require(
-                groupSignature_.length == bClaims_.length && groupSignature_.length >= 1,
-                string(abi.encodePacked(SnapshotsErrorCodes.SNAPSHOT_MIGRATION_INPUT_DATA_MISMATCH))
-            );
+            if (_epoch != 0) {
+                revert SnapshotsErrors.MigrationNotAllowedAtCurrentEpoch();
+            }
+            if (!(groupSignature_.length == bClaims_.length && groupSignature_.length >= 1)) {
```
Comment (vtleonardo) [inline bridge/contracts/Snapshots.sol:163]: What do you think about reverting the boolean expression for more clarity? 
```
groupSignature_.length != bClaims_.length || groupSignature_.length < 1
```

Comment (keirongulrajani) [reply bridge/contracts/Snapshots.sol:163]: There's quite a few instances where i wrapped the parameters of the old `require` statements and checked for the reverse given that i was converting to `if` statements and I wanted to ensure the logic wasn't broken. I'll go through the instances of this and refactor for readability 👍

File: bridge/contracts/Snapshots.sol [hunk 4]
```diff
@@ -53,55 +53,55 @@ contract Snapshots is Initializable, SnapshotsStorage, ISnapshots {
...
                 .extractSigGroup(groupSignature_, 0);
 
-            require(
-                keccak256(abi.encodePacked(masterPublicKey)) ==
-                    IETHDKG(_ethdkgAddress()).getMasterPublicKeyHash(),
-                string(abi.encodePacked(SnapshotsErrorCodes.SNAPSHOT_WRONG_MASTER_PUBLIC_KEY))
-            );
+            if (
+                keccak256(abi.encodePacked(masterPublicKey)) !=
+                IETHDKG(_ethdkgAddress()).getMasterPublicKeyHash()
```
Comment (vtleonardo) [inline bridge/contracts/Snapshots.sol:78]: Add the both hashes to the error

File: bridge/contracts/ValidatorPool.sol [hunk 5]
```diff
@@ -33,36 +33,32 @@ contract ValidatorPool is
...
 
     constructor() ValidatorPoolStorage() {}
 
     receive() external payable {
-        require(
-            msg.sender == _validatorStakingAddress() || msg.sender == _publicStakingAddress(),
-            string(abi.encodePacked(ValidatorPoolErrorCodes.VALIDATORPOOL_ONLY_CONTRACTS_ALLOWED))
-        );
+        if (msg.sender != _validatorStakingAddress() && msg.sender != _publicStakingAddress()) {
+            revert ValidatorPoolErrors.OnlyNFTContractsAllowed();
```
Comment (vtleonardo) [inline bridge/contracts/ValidatorPool.sol:60]: Only staking contracts allowed?

File: bridge/contracts/ValidatorPool.sol [hunk 6]
```diff
@@ -466,16 +427,12 @@ contract ValidatorPool is
...
-            _validators.length() <= _maxNumValidators,
-            string(abi.encodePacked(ValidatorPoolErrorCodes.VALIDATORPOOL_MAX_VALIDATORS_MET))
-        );
-        require(
-            !_isAccusable(validator_),
-            string(
-                abi.encodePacked(ValidatorPoolErrorCodes.VALIDATORPOOL_ADDRESS_ALREADY_VALIDATOR)
-            )
-        );
+        if (_validators.length() > _maxNumValidators) {
```
Comment (vtleonardo) [inline bridge/contracts/ValidatorPool.sol:430]: There's a bug in the current implementation, after reviwing the old code. Can you change this to be >= ?

File: bridge/contracts/ValidatorPool.sol [hunk 7]
```diff
@@ -116,14 +112,13 @@ contract ValidatorPool is
...
-            string(
-                abi.encodePacked(ValidatorPoolErrorCodes.VALIDATORPOOL_MIN_BLOCK_INTERVAL_NOT_MET)
-            )
-        );
+        if (
+            block.number <=
+            ISnapshots(_snapshotsAddress()).getCommittedHeightFromLatestSnapshot() +
+                MAX_INTERVAL_WITHOUT_SNAPSHOTS
+        ) {
+            revert ValidatorPoolErrors.MinimumBlockIntervalNotMet();
```
Comment (z-j-lin) [inline bridge/contracts/ValidatorPool.sol:120]: can you add .getCommittedHeightFromLatestSnapshot() + MAX_INTERVAL_WITHOUT_SNAPSHOTS, and block.number to the variables in the error. for debugging purposes

File: bridge/contracts/libraries/ethdkg/ETHDKGAccusations.sol [hunk 8]
```diff
@@ -47,44 +46,39 @@ contract ETHDKGAccusations is ETHDKGStorage, IETHDKGEvents, ETHDKGUtils {
...
-                ((block.number >= _phaseStartBlock + _phaseLength) &&
-                    (block.number < _phaseStartBlock + 2 * _phaseLength)),
-            string(abi.encodePacked(ETHDKGErrorCodes.ETHDKG_NOT_IN_POST_SHARED_DISTRIBUTION_PHASE))
-        );
+        if (
+            _ethdkgPhase != Phase.ShareDistribution ||
+            block.number < _phaseStartBlock + _phaseLength ||
+            block.number >= _phaseStartBlock + 2 * _phaseLength
+        ) {
+            revert ETHDKGErrors.NotInPostSharedDistributionPhase(_ethdkgPhase);
```
Comment (vtleonardo) [inline bridge/contracts/libraries/ethdkg/ETHDKGAccusations.sol:54]: Same suggestion as above to add more info.

File: bridge/contracts/libraries/ethdkg/ETHDKGAccusations.sol [hunk 9]
```diff
@@ -16,27 +16,26 @@ contract ETHDKGAccusations is ETHDKGStorage, IETHDKGEvents, ETHDKGUtils {
...
-                ((block.number >= _phaseStartBlock + _phaseLength) &&
-                    (block.number < _phaseStartBlock + 2 * _phaseLength)),
-            string(abi.encodePacked(ETHDKGErrorCodes.ETHDKG_NOT_IN_POST_REGISTRATION_PHASE))
-        );
+        if (
+            _ethdkgPhase != Phase.RegistrationOpen ||
+            block.number < _phaseStartBlock + _phaseLength ||
+            block.number >= _phaseStartBlock + 2 * _phaseLength
+        ) {
+            revert ETHDKGErrors.ETHDKGNotInPostRegistrationAccusationPhase(_ethdkgPhase);
```
[the remark this answers is not in the document]
Comment (vtleonardo) [reply bridge/contracts/libraries/ethdkg/ETHDKGAccusations.sol:24]: This should be a follow-up ticket due to the complexity.

File: bridge/contracts/libraries/ethdkg/ETHDKGAccusations.sol [hunk 10]
```diff
@@ -101,69 +95,71 @@ contract ETHDKGAccusations is ETHDKGStorage, IETHDKGEvents, ETHDKGUtils {
         uint256[2] memory sharedKeyCorrectnessProof
     ) external {
         // We should allow accusation, even if some of the participants didn't participate
-        require(
-            (_ethdkgPhase == Phase.DisputeShareDistribution &&
+        if (
+            !(_ethdkgPhase == Phase.DisputeShareDistribution &&
```
Comment (vtleonardo) [inline bridge/contracts/libraries/ethdkg/ETHDKGAccusations.sol:99]: Let's revert the boolean logic here for more clarity.

File: bridge/contracts/libraries/parsers/MerkleProofParserLibrary.sol [hunk 11]
```diff
@@ -2,8 +2,8 @@
 pragma solidity ^0.8.11;
 
 import {
-    MerkleProofParserLibraryErrorCodes
-} from "contracts/libraries/errorCodes/MerkleProofParserLibraryErrorCodes.sol";
+    MerkleProofParserLibraryErrors
```
Comment (vtleonardo) [inline bridge/contracts/libraries/parsers/MerkleProofParserLibrary.sol:5]: Let's remove the import {X}  from here as well.

File: bridge/contracts/utils/CustomEnumerableMaps.sol [hunk 12]
```diff
@@ -2,8 +2,8 @@
 pragma solidity ^0.8.11;
 
 import {
-    CustomEnumerableMapsErrorCodes
-} from "contracts/libraries/errorCodes/CustomEnumerableMapsErrorCodes.sol";
+    CustomEnumerableMapsErrors
```
Comment (vtleonardo) [inline bridge/contracts/utils/CustomEnumerableMaps.sol:5]: Remove import from logic from here

File: bridge/contracts/utils/ImmutableAuth.sol [hunk 13]
```diff
@@ -3,16 +3,15 @@
...
     address private immutable _factory;
+    error OnlyFactory(address sender);
 
     modifier onlyFactory() {
-        require(
-            msg.sender == _factory,
-            string(abi.encodePacked(ImmutableAuthErrorCodes.IMMUTEABLEAUTH_ONLY_FACTORY))
-        );
+        if (msg.sender != _factory) {
+            revert OnlyFactory(msg.sender);
```
Comment (vtleonardo) [inline bridge/contracts/utils/ImmutableAuth.sol:13]: Can we pass the expected address on these errors as well?

**Review by vtleonardo — APPROVED** [review]

**Comment by vtleonardo** [comment]
All contracts and unit tests reviewed. Integration tests passed:
- [x] ETHDKG finished without errors
- [x] Snapshots happened
- [x] Fast sync and normal sync worked
- [x] Deposit worked
Custom errors are being displayed on the golang code!

**Review by nelsonhp — APPROVED** [review]
Insane amount of logic inversion. Well done - in future we should break this up more.

**Review by z-j-lin — APPROVED** [review]

**Pull request merged**